### PR TITLE
Address guvectorize too slow for cuda target

### PR DESCRIPTION
**Pull request by testhound**
This pull request address: https://github.com/numba/numba/issues/6159. It addresses the issues by detecting when the blocksize for a kernel would be < 2 * Number of SM's. I also added a environment variable called NUMBA_CUDA_LOW_OCCUPANCY_WARNINGS that is enabled by default. I had to modify the the class CUDATestCase to disable the warning as many of the test cases trigger this warning.

### CODE REVIEW

**Comment by sklam** [comment]
Started smoketest: `numba_smoketest_cuda_148`

File: docs/source/reference/envvars.rst
Comment (gmarkall) [inline docs/source/reference/envvars.rst:472]: The error is generated when the grid size is too small, not the block size. It's probably worth mentioning that the heuristic used here is `(grid size) < 2 * (number of SMs)`, and that the absence of the warning doesn't necessarily imply a good grid size - larger multiples of the SM count may be more efficient.

File: numba/cuda/tests/cudapy/test_gufunc.py
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_gufunc.py:190]: I think this could do with a comment explaining that the test creates an inefficient use of the GPU because the inputs are all mapped onto a single thread in a single block.

File: docs/source/reference/envvars.rst
Comment (gmarkall) [reply docs/source/reference/envvars.rst:472]: Also it should be mentioned that this is on by default.

File: numba/cuda/compiler.py
Comment (gmarkall) [inline numba/cuda/compiler.py:795]: I think `will generate inefficient kernel code` is a bit confusing - how about phrasing the warning as `Grid size (<size>) less than 2 * SM count (2 * (<SM count>) = (<SM count * 2>) will likely result in GPU underutilization due to low occupancy`?

**Review by gmarkall — CHANGES_REQUESTED** [review]
This is looking promising; there are some comments on the diff suggesting some changes.

It's good that there is a test ensuring this fires for inefficient uses of `@guvectorize` functions. I think there should also be a test of this using a normal kernel launch - I can't see a really good place to put this new test, but either in `numba/cuda/tests/cudapy/test_errors.py` or in a new file `test_warnings.py`.

**Comment by testhound** [comment]
@gmarkall I believe all your review comments have been incorporated.

File: numba/core/config.py
Comment (stuartarchibald) [inline numba/core/config.py:159]:
```suggestion
        CUDA_LOW_OCCUPANCY_WARNINGS = _readenv(
```
All other CUDA related config vars start with `CUDA_` or have `CUDA` in their name, suggest calling this something like the above.

File: numba/cuda/testing.py
Comment (stuartarchibald) [inline numba/cuda/testing.py:32]:
```suggestion
        # Disable warnings about low gpu utilization in the test suite
```

File: numba/cuda/compiler.py
Comment (stuartarchibald) [inline numba/cuda/compiler.py:796]:
```suggestion
                       "will likely result in GPU under utilization due"
```

File: numba/core/config.py
Comment (stuartarchibald) [inline numba/core/config.py:158]:
```suggestion
        # under utilize the GPU due to low occupancy. On by default.
```

File: docs/source/reference/envvars.rst
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:473]:
```suggestion
   streaming multiprocessors (SM). This option is on by default (default value is 1).
```

File: docs/source/reference/envvars.rst
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:475]:
```suggestion
   The heuristic checked is whether ``gridsize < 2 * (number of SMs)``. NOTE: The absence of
```

File: docs/source/reference/envvars.rst
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:478]: Does lead to a noticeable compilation speed gain?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this is great feedback to give users, it's good to see this implemented. I've a few minor comments in addition to @gmarkall's review, once resolved looks good to merge. Thanks again!

File: docs/source/reference/envvars.rst
Comment (gmarkall) [reply docs/source/reference/envvars.rst:478]: This is for launch speed rather than compilation speed. In the past there have been performance issues due to excessive getting of contexts, e.g.: https://github.com/numba/numba/pull/5189 - so I think if we add a get of the context in the launch pipeline, it's good to have a way around it as well.

**Comment by gmarkall** [comment]
> @gmarkall I believe all your review comments have been incorporated.

Thanks for looking at these. I think they are all addressed, except this one (from the initial review):

> I think there should also be a test of this using a normal kernel launch - I can't see a really good place to put this new test, but either in numba/cuda/tests/cudapy/test_errors.py or in a new file test_warnings.py.

File: numba/core/config.py
Comment (testhound) [reply numba/core/config.py]: Done.

File: numba/cuda/testing.py
Comment (testhound) [reply numba/cuda/testing.py]: Done

File: numba/cuda/compiler.py
Comment (testhound) [reply numba/cuda/compiler.py]: Done.

File: numba/core/config.py
Comment (testhound) [reply numba/core/config.py]: Done.

File: docs/source/reference/envvars.rst
Comment (testhound) [reply docs/source/reference/envvars.rst]: Done.

File: docs/source/reference/envvars.rst
Comment (testhound) [reply docs/source/reference/envvars.rst]: Done.

**Comment by testhound** [comment]
> > @gmarkall I believe all your review comments have been incorporated.
> 
> Thanks for looking at these. I think they are all addressed, except this one (from the initial review):
> 
> > I think there should also be a test of this using a normal kernel launch - I can't see a really good place to put this new test, but either in numba/cuda/tests/cudapy/test_errors.py or in a new file test_warnings.py.

@gmarkall I addressed this with the function 'test_efficient_guvectorize' in the file test_gufunc.py.

**Comment by gmarkall** [comment]
Sorry, I forgot to respond earlier - I mean a test where you call a kernel like `f[1,1](arg)` and check the message is produced (and one that launches with better occupancy and checks that there is no warning). I do think `test_efficient_guvectorize` is a good addition too though.

File: numba/cuda/tests/cudapy/test_gufunc.py
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_gufunc.py:191]: (Note: this is resolved, but I can't resolve the comment because I don't have write access to this repo)

File: numba/cuda/tests/cudapy/test_gufunc.py
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_gufunc.py:200]: Looking at the performance of the kernel test led me to also look into how long the guvectorize tests take - I think the data sizes for this and the test below are too large for unit tests, and they take about 5 seconds to run - can we reduce the size of the data please - e.g.:

```suggestion
        a = np.random.rand(1024 * 32).astype('float32')
        b = np.random.rand(1024 * 32).astype('float32')
```

File: numba/cuda/tests/cudapy/test_gufunc.py
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_gufunc.py:223]: Also to reduce the data size here please:

```suggestion
        a = np.random.rand(524288 * 2).astype('float32').reshape((524288, 2))
        b = np.random.rand(524288 * 2).astype('float32').reshape((524288, 2))
```

The leading dimension of 524288 is needed to get a big enough grid size for a large GPU, so I reduced the trailing dimension down to 2 to keep the data size as small as possible.

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the updated tests! I have a few comments regarding the kernel launch test - rather than annotating lines on the diff, I'll try and summarise here:

- The tests are quite complex compared to what's needed, resulting in the runtime being too long and the purpose of the tests being slightly obscured - the runtime on my machine is 40 seconds for `test_warnings.py`. Since we're only trying to test whether the warning is generated for a low occupancy configuration, we don't need the kernel to do anything or create any arrays to pass into it.
- I think that even if the test runtime is reduced a lot, the test still doesn't work in the CUDA simulator, because it doesn't generate performance warnings (and we shouldn't try to make it do so).
- The names of the tests are a bit vague about what is really tested - `inefficient_kernel` implies that the problem is inefficient code in the kernel, rather than the launch configuration.
- A minor point: I try to not include `CUDA` in the names of CUDA test classes, because it's redundant in the CUDA test suite - the main issue I have with it is having to type more when running individual classes from the command line :-)

I think the above can be resolved with the following implementation of the test class:

```python
@skip_on_cudasim('cudasim does not raise performance warnings')
class TestWarnings(CUDATestCase):
    def test_inefficient_launch_configuration(self):
        @cuda.jit
        def kernel():
            pass

        with override_config('CUDA_LOW_OCCUPANCY_WARNINGS', 1):
            with warnings.catch_warnings(record=True) as w:
                kernel[1, 1]()

        self.assertEqual(w[0].category, NumbaPerformanceWarning)
        self.assertIn('Grid size', str(w[0].message))
        self.assertIn('2 * SM count', str(w[0].message))

    def test_efficient_launch_configuration(self):
        @cuda.jit
        def kernel():
            pass

        with override_config('CUDA_LOW_OCCUPANCY_WARNINGS', 1):
            with warnings.catch_warnings(record=True) as w:
                kernel[256, 256]()

        self.assertEqual(len(w), 0)
```

There are also a couple of comments / questions on the diff to look at - one is related to the way `setUp` works, which turned out to be more complex than I first imagined it would be.

File: numba/cuda/testing.py
Comment (testhound) [reply numba/cuda/testing.py:28]: @gmarkall good catch on setUpClass. I don't think however, removing the __init __ method and it's implementation is correct. If you remove __init__ then 'self._low_occupancy_warnings" will not be created as a instance variable.

**Comment by testhound** [comment]
@gmarkall should the tests in test_gufunc be skipped for the CUDA simulator (since no warnings are generated in the simulator), similar to the new ones in test_warning.py?

File: numba/cuda/testing.py
Comment (gmarkall) [reply numba/cuda/testing.py:28]: It will be created during the `setUp()` method - it doesn't need to be "declared" in `__init__()`.

**Comment by gmarkall** [comment]
> @gmarkall should the tests in test_gufunc be skipped for the CUDA simulator (since no warnings are generated in the simulator), similar to the new ones in test_warning.py?

The whole test class there is skipped, so there's no need for additional decoration of the test methods you've added.

**Review by gmarkall — DISMISSED** [review]
Many thanks for the updates! This looks good to me now, and passes tests locally.

@Stuartarchibald / @esc could this have a buildfarm run please? The test fail showing presently is something broken in the main branch.

**Comment by sklam** [comment]
BFID `numba_smoketest_cuda_yaml_60`

File: docs/source/reference/envvars.rst
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:473]:
```suggestion
   streaming multiprocessors (SM). This option is on by default (default value is 1).
```

File: docs/source/reference/envvars.rst
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:478]:
```suggestion
   heuristic needs to check the number of SMs available on the device in the
   current context.
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, couple of tiny doc change suggestions else looks good. Thanks for reviewing @gmarkall .

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_60`.

**Comment by stuartarchibald** [comment]
> BFID `numba_smoketest_cuda_yaml_60`

This passed.

File: docs/source/reference/envvars.rst
Comment (testhound) [reply docs/source/reference/envvars.rst]: @stuartarchibald change committed.

File: docs/source/reference/envvars.rst
Comment (testhound) [reply docs/source/reference/envvars.rst]: @stuartarchibald change committed.

**Comment by sklam** [comment]
`numba.tests.test_config.TestConfig.test_config` is failing.

**Comment by testhound** [comment]
> `numba.tests.test_config.TestConfig.test_config` is failing.

Yes I see that but it's not happening on all platforms or on my local machine.

**Comment by gmarkall** [comment]
> > `numba.tests.test_config.TestConfig.test_config` is failing.
> 
> Yes I see that but it's not happening on all platforms or on my local machine.

It does seem repeatable though. Since the diff is not visible, to debug this I'd push a commit that sets `self.maxDiff` to something greater than 2109 in `numba.tests.test_config.TestConfig` as suggested in the error log to see what the difference is.

**Comment by testhound** [comment]
> > > `numba.tests.test_config.TestConfig.test_config` is failing.
> > 
> > 
> > Yes I see that but it's not happening on all platforms or on my local machine.
> 
> It does seem repeatable though. Since the diff is not visible, to debug this I'd push a commit that sets `self.maxDiff` to something greater than 2109 in `numba.tests.test_config.TestConfig` as suggested in the error log to see what the difference is.

@gmarkall okay done.

File: numba/cuda/compiler.py
Comment (gmarkall) [inline numba/cuda/compiler.py:834]: I just noticed there's a space missing here so it ends up printing "dueto low occupancy."

```suggestion
                       "will likely result in GPU under utilization due "
                       "to low occupancy.")
```

**Comment by gmarkall** [comment]
The log shows the difference in configurations between expected and actual now:

```
-  'CUDA_LOW_OCCUPANCY_WARNINGS': 0,
?                                 ^

+  'CUDA_LOW_OCCUPANCY_WARNINGS': 1,
?                                 ^
```

**Comment by gmarkall** [comment]
It seems as though the conditions creating this are (in order of occurrence):

1. A `CUDATestCase` class is setup up, setting `config.CUDA_LOW_OCCUPANCY_WARNINGS` to `0`
2. `TestConfig.test_config` runs
3. The teardown for the `CUDATestCase` happens (or it never happens)

I'm not immediately  sure why this is - I'd expected the tests to be run one class at a time but this expectation may not be correct.

File: numba/cuda/testing.py
Comment (sklam) [inline numba/cuda/testing.py:38]: Why is this needed? 

If we need to guarantee that this `setUp` is called first, it's better for all the subclasses to have:

```python
def setUp(self):
    super().setUp()
    .... # local setup
```

Comment (testhound) [reply numba/cuda/testing.py:38]: @sklam is your suggestion that the setupClass method is not required and that a call to the base class setup is recommended as shown?

Comment (sklam) [reply numba/cuda/testing.py:38]: Yes, give that a try.  See if avoiding the `setUpClass` fixes the test failure.

File: numba/cuda/testing.py
Comment (gmarkall) [inline numba/cuda/testing.py:43]: I don't think this is necessary, because `ContextResettingTestCase.setUp()` isn't doing anything.

File: numba/cuda/tests/cudadrv/test_cuda_devicerecord.py
Comment (gmarkall) [inline numba/cuda/tests/cudadrv/test_cuda_devicerecord.py:89]: This is now calling `TestCudaDeviceRecord.setUp()`, which wasn't the case previously - the idea here is that `TestCudaDeviceRecordWithRecord` has a distinct `setUp` method. So I think this needs to be:

```suggestion
        CUDATestCase.setUp()
```

in order to bypass `TestCudaDeviceRecord.setUp()`.

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the changes. This passes locally for me with the CUDA target.

I see that the fix removes the use of `setUpClass` and "manually" invokes `setUp` where required in each derived class. I'm not sure I understand what the problem with `setUpClass` was though - is there some interleaving of the running of tests from different classes that caused the issue? (@sklam)?

There are a couple of small points on the diff - one is more "stylistic" and the other is where the behavior of a `setUp` appears to be changed.

**Comment by gmarkall** [comment]
> `setUpClass` is probably running as soon as the test is loaded. Maybe it should be a `setUp` to surround just the test method.

If this is the case, it seems to be at odds with the documentation (https://docs.python.org/3/library/unittest.html#class-and-module-fixtures), which states:

> When the test suite encounters a test from a new class then tearDownClass() from the previous class (if there is one) is called, followed by setUpClass() from the new class.

Do you see another interpretation of what is happening?

**Review by gmarkall — APPROVED** [review]
Thanks for the updates - looks good to me!

@stuartarchibald Do you want another buildfarm run, or are you happy with this as-is?

**Comment by gmarkall** [comment]
... Actually - if a buildfarm run can be squeezed in just to be on the safe side, that would be good - is that possible? (@stuartarchibald / @esc)

File: numba/cuda/tests/cudadrv/test_cuda_devicerecord.py
Comment (sklam) [inline numba/cuda/tests/cudadrv/test_cuda_devicerecord.py:89]: Missing `self`. Should be `super().setUp()`

**Review by sklam — CHANGES_REQUESTED** [review]
Tests are failing because of the following:

File: numba/cuda/tests/cudadrv/test_cuda_devicerecord.py
Comment (gmarkall) [reply numba/cuda/tests/cudadrv/test_cuda_devicerecord.py:89]: This will call `TestCudaDeviceRecord.setUp()`, which is what I was trying to avoid with the suggestion to use `CUDATestCase.setUp()` instead.

Comment (sklam) [reply numba/cuda/tests/cudadrv/test_cuda_devicerecord.py:89]: In that case, it needs to be `CUDATestCase.setUp(self)`

Comment (sklam) [reply numba/cuda/tests/cudadrv/test_cuda_devicerecord.py:89]: Example of errors:

```pytb
======================================================================
ERROR: test_auto_device (numba.cuda.tests.cudadrv.test_cuda_devicerecord.TestCudaDeviceRecordWithRecord)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/siu/dev/numba/numba/cuda/tests/cudadrv/test_cuda_devicerecord.py", line 89, in setUp
    CUDATestCase.setUp()
TypeError: setUp() missing 1 required positional argument: 'self'

======================================================================
ERROR: test_device_record_copy (numba.cuda.tests.cudadrv.test_cuda_devicerecord.TestCudaDeviceRecordWithRecord)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/siu/dev/numba/numba/cuda/tests/cudadrv/test_cuda_devicerecord.py", line 89, in setUp
    CUDATestCase.setUp()
TypeError: setUp() missing 1 required positional argument: 'self'

======================================================================
ERROR: test_device_record_interface (numba.cuda.tests.cudadrv.test_cuda_devicerecord.TestCudaDeviceRecordWithRecord)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/siu/dev/numba/numba/cuda/tests/cudadrv/test_cuda_devicerecord.py", line 89, in setUp
    CUDATestCase.setUp()
TypeError: setUp() missing 1 required positional argument: 'self'

======================================================================
ERROR: test_from_record_like (numba.cuda.tests.cudadrv.test_cuda_devicerecord.TestCudaDeviceRecordWithRecord)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/siu/dev/numba/numba/cuda/tests/cudadrv/test_cuda_devicerecord.py", line 89, in setUp
    CUDATestCase.setUp()
TypeError: setUp() missing 1 required positional argument: 'self'

----------------------------------------------------------------------
```

Comment (gmarkall) [reply numba/cuda/tests/cudadrv/test_cuda_devicerecord.py]: Apologies for missing `self` in my earlier suggestion - I thought I tested my suggestions locally, but I must have done something not quite right!

**Review by gmarkall — APPROVED** [review]
LGTM!

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes @testhound.

Thanks for reviewing @gmarkall.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_68`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_68`.

Passed

**Review by sklam — APPROVED** [review]

**Pull request merged**
